### ISSUE TITLE
Error in annotation by batch

### ISSUE THREAD

**Issue by GWheelerEB** [ChristofferFlensburg/superFreq#39]
I am running into an issue at the VariantAnnotation step, which appears to be at the step where indels are processed and a series of Ns added. The error is as follows:

`Starting differential coverage analysis by sample.
Loading saved differential coverage results.
Loaded saved differential coverage results
Plotting volcanoes to /igm/home/glw001/Clonal/Test5/Plots/Proband/volcanoes/..done!
Loading final version of combined variants.
Average variant loss is 0.04953373
Running VariantAnnotation.
Loading annotation dump...done.
Splitting up 3049 variants for parallelisation into 1 batches.
Setting up data bases..done.
Running annotation by batch.Error in rep("N", n + 1) : invalid 'times' argument
Calls: superFreq ... sapply -> sapply -> lapply -> FUN -> do.call -> as.list
In addition: Warning message:
In if (any(grepl("-", variant))) { :
  the condition has length > 1 and only the first element will be used
Execution halted`

I've dug around in the code a bit to try to remedy this, but honestly have not been able to make too much progress. 'variants' contains a list with a mix of SNPs and indels. 'nDel' contains either numbers (for deletions) or NAs (for others). Changes I've attempted have caused errors in downstream steps; for example, replacing the NAs with zeros results in this error:

`Error in .nextMethod(value = value, x = x) :
  'value' must be a DNAStringSet object (or coercible to a DNAStringSet object)
Calls: superFreq ... normalizeSingleBracketReplacementValue -> callNextMethod -> .nextMethod
In addition: Warning message:
In if (any(grepl("-", variant))) { :
  the condition has length > 1 and only the first element will be used
Execution halted`

As a point of clarification, I am not concerned about whether or not the reference or variant for an indel is composed of Ns (which I see is already a posted issue). I'm just trying to get past this step at all.

**Comment by ChristofferFlensburg** [comment]
Hi Greg!

Thanks for raising the issue! I'm not immediately sure what the cause is though.

I had a look at the function and it seems like it's behaving as expected for a few tests I ran. The sanity checking of the input isn't great though, so I suspect that it's an upstreams issues that manifests in this function.

I know you can't share data, but maybe you can have a quick look for me to confirm? In the `variants` object, can you check if the `variant' column (so `variants$variant`) is ok? Maybe run a table() and make sure it's only ATCGs and -[number] in that column. You can load from the latest savepoint which is the allVariants[preVEP].Rdata in the R directory.

Without reproduce the error, it's tricky for me to solve. :/ But I am adding in more check in that part of the code now, so worst case we can rerun with those and hope for more informative warnings/errors.

Also if you can share the .log file (in the R directory), it might have clues to what went wrong. The log file doesn't contain any sensitive variant information, but it does contain the meta data you input to superFreq, with the sample and indiviudal names you set, as well as your paths to the input and output files.

cheers,
/Christoffer

**Comment by ChristofferFlensburg** [comment]
Hey, I added in a warning message where I think things might have gone wrong. If you can run again with version 1.2.6, there might be a more informative warning. Sorry for the mess here, it's so difficult to figure out what is going on when no one can share the data that causes issues.

**Comment by GWheelerEB** [comment]
I've run again with the new version and get a different error, though I'm not sure if it's more informative. It is as follows:
`Splitting up 3049 variants for parallelisation into 1 batches.
Setting up data bases..done.
Running annotation by batch.Error in rep("N", n + 1) : invalid 'times' argument
Calls: superFreq ... sapply -> sapply -> lapply -> FUN -> do.call -> as.list
In addition: Warning messages:
1: In if (any(grepl("-", variant))) { :
  the condition has length > 1 and only the first element will be used
2: In convertVariantsToVCF(list(REF, ALT, q$x, q$x)) :
  NA nDel in convertVariantsToVCF. Causing variants are:
Execution halted
`
I've checked the variants list, and I saw A, T, G, C, negative integers for deletions, and "+" plus strings of bases for insertions. From your previous response, it seems like maybe the insertions are not supposed to be in there? I've also attached the log file, which doesn't contain any identifying information but also just looks to have stdout, not any additional error information. As for the variants themselves, I'll have to get additional confirmation of how much I can share.

[runtimeTracking.log](https://github.com/ChristofferFlensburg/superFreq/files/3167579/runtimeTracking.log)

**Comment by ChristofferFlensburg** [comment]
Hey!

Thanks for following up.

Hmm, so the warning and errors come from these lines:
``` R
    deletions = grepl('-', variant)
    nDel = as.numeric(gsub('-', '', variant[deletions]))
    if ( any(is.na(nDel)) ) warning('NA nDel in convertVariantsToVCF. Causing variants are: ', paste(variant[deletions][is.na(nDel)], collapse=','))
    reference[deletions] = sapply(nDel, function(n) do.call(paste0, as.list(rep('N', n+1))))
```
Where `variant` should be the entries you just looked up, and it seems like those are in order. The `+ACTATG` are expected, I just forgot to mention them. The deletions extract only variants with a `-` in them. That kind of input string should work fine through that code, or at least it does in testing, so probably there is some bad `variant` input.

However, right now I can't even figure out what `variant` input would trigger the `any(is.na(nDel))` condition, but return nothing (or an emtpy strings) from the `paste(variant[deletions][is.na(nDel)], collapse=',')` in the warning...

And the more I think about it, I am equally confused by the warning
```
In if (any(grepl("-", variant))) { : the condition has length > 1 and only the first element will be used
```
because `any()` should only be able to return `TRUE', 'FALSE', or 'NA' no matter what you give it, neither of which has length more than 1 and triggers this warning.

Quite a puzzle, but I'd like this solved! Maybe you can give me your sessionInfo(), to check versions? Do you have the possibility to run this interactively in debug mode (`options(error=utils::recover)` should do) otherwise, and see what `variant', 'deletions' and 'nDel' is? If not, I can add some more warnings with that kindof output. :P

**Comment by GWheelerEB** [comment]
Here's the outputs of sessionInfo() for the environment I'm running in. For a bit of further info, I am running on an AWS SGE node. I've been executing this by starting an interactive 'qsub' job and executing an R-script I have set up.

`R version 3.6.0 (2019-04-26)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS:   /igm/apps/R/R-3.6.0_install/lib64/R/lib/libRblas.so
LAPACK: /igm/apps/R/R-3.6.0_install/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
[1] compiler_3.6.0
`

Setting options(error=recover) and running again, I get these frames to access:
` 1: superFreq(metaDataFile, captureRegions = captureRegionsFile, normalDirector
 2: analyse(inputFiles = inputFiles, outputDirectories = outputDirectories, set
 3: annotateSomaticQs(variants$variants, genome = genome, resourceDirectory = r
 4: annotateQ(somQ, genome = genome, resourceDirectory = resourceDirectory, ref
 5: lapply(qList, function(q) {
    catLog(".")
    vcf = superFreq:::qToVCF(q,
 6: lapply(qList, function(q) {
    catLog(".")
    vcf = superFreq:::qToVCF(q,
 7: FUN(X[[i]], ...)
 8: superFreq:::qToVCF(q, genome = genome, seqlengths = seqlengths(txdb), addCh
 9: convertVariantsToVCF(list(REF, ALT, q$x, q$x))
10: sapply(nDel, function(n) do.call(paste0, as.list(rep("N", n + 1))))
11: sapply(nDel, function(n) do.call(paste0, as.list(rep("N", n + 1))))
12: lapply(X = X, FUN = FUN, ...)
13: FUN(X[[i]], ...)
14: do.call(paste0, as.list(rep("N", n + 1)))
15: as.list(rep("N", n + 1))
`

Looking at 15, 'variant' gives this output:

`[[1]] -1
[[2]] -1
[[3]] +A
[[4]] G
[[5]] T
[[6]] G
[[7]] T
[[8]] G
[[9]] G
[[10]] T
...
<3039 more elements>
`
I've exported this and manually looked through everything, but it seems fine.

'deletions' gives this:
`LogicalList of length 3049
[[1]] TRUE
[[2]] TRUE
[[3]] FALSE
[[4]] FALSE
[[5]] FALSE
[[6]] FALSE
[[7]] FALSE
[[8]] FALSE
[[9]] FALSE
[[10]] FALSE
...
<3039 more elements>
`

I wrote this out as well and lined it up with 'variants'. All the deletions are TRUE; everything else is false; all fields have a value.

nDel is an object of type 'numeric' (not a list, if that matters) with 3049 values. Every field is either an integer or 'NA'. I lined it up with the other data; every deletion has an integer value (all appeared to be correct), and everything else had an 'NA'.

**Comment by ChristofferFlensburg** [comment]
Thanks for all the work!

I think I figured it out now. 85% certain! 👍 
`deletions` should be of class logical, not a CompressedLogicalList. The call to `grepl` with the CharacterList input has returned a logical vector in the versions of IRanges I've been using and testing on, but not for you... Must be a different version of IRanges maybe? Current version is 2.16.0, I guess you have an older one? Otherwise I'll be confused again. :P This causes downstream base function calls to instead invoke overloaded IRanges version of the function that takes CompressedLogicalList as input, which apparently behaves differently enough to crash superFreq. And that's why I've been so confused, because what I thought were calls to base functions returned values that the base functions shouldn't be able to return, while it was actually the overloaded IRanges functions that were called...

If you want to confirm, try this:
``` R
library(IRanges)
packageVersion('IRanges') #I just updated to 2.16.0
variant = CharacterList(as.list(c('A', 'T', '+A', '-5')))
deletions = grepl('-', variant)
class(deletions) #This returns 'logical' for me, your version probably 'CompressedLogicalList'?
```

Anyway, instead of fiddling with figuring out which versions of IRanges overload `grepl` in different ways, I just bypassed the problem by explicitly converting `variant` to a plain character class, and everything will stick to base version of functions. Then convert back to CharacterList when I call the IRanges functions that need it. I will do the test runs overnight and push 1.2.7 online tomorrow (if tests pass).

If you're impatient, you can try to update IRanges, test the above code snippet and run again on current version if it returns logical. May as well update all the dependencies while you're at it I guess.

**Comment by GWheelerEB** [comment]
I've just confirmed that my 'IRanges' is version 2.**18**.0, which appears to be the current version hosted on Bioconductor. Running the provided code, it does return that deletions is a "CompressedLogicalList".

**Comment by ChristofferFlensburg** [comment]
Oh, wow, and here I thought that `devtools::install` also updates bioconductor packages, but seems like it doesn't! :o That's embarrassing, sorry about that. We live, we learn.

Well I updated all my packages now, and also reproduce your results in 2.18.0 IRanges. Made the bug fix in the code, and at least one of the tests passed the annotation where you got the crash, so I pushed version 1.2.7 online just now. Trusting the rest of the tests will go through as well, or we will have 1.2.8 soon as well! :P So reinstall and rerun, and it should work for you. Hopefully.

Thanks for staying with me through this journey, and I hope you get good results! :)
